### PR TITLE
OP-TEE Benchmark

### PR DESCRIPTION
**Pull request by igoropaniuk**
Related PRs:

Benchmark app:
https://github.com/linaro-swg/optee_benchmark/pull/1

Changes in OP-TEE
https://github.com/linaro-swg/linux/pull/38
https://github.com/OP-TEE/build/pull/124
https://github.com/OP-TEE/optee_os/pull/1365
https://github.com/OP-TEE/optee_client/pull/79

### CODE REVIEW

File: libteec/src/tee_bench_aux.c [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <tee_client_api.h>
+#include <unistd.h>
+
+struct tee_time_buf *bench_ts_buf;
+struct statistics *bench_stat_buf;
+
+#ifdef CFG_TEE_BENCHMARK
+static const TEEC_UUID sta_benchmark_uuid = BENCHMARK_STA_UUID;
```
Comment (etienne-lms) [inline libteec/src/tee_bench_aux.c]: minor: rather user `pta_xxx` than `sta_xxx`. "static TA" is a deprecated reference.
here and there.

File: libteec/src/tee_bench_aux.c [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+}
+
+/* Take new sample into account (Knuth/Welford algorithm) */
+static void update_stats(struct statistics *s, uint64_t t)
+{
+	double x = (double)t;
+	double delta = x - s->m;
+
+	s->n++;
+	s->m += delta/s->n;
```
Comment (etienne-lms) [inline libteec/src/tee_bench_aux.c]: minor: spaces around `/`

File: libteec/src/tee_bench_aux.c [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+	off_t offset = 0;
+	off_t page_addr;
+
+	intptr_t *hw_addr = (intptr_t *)paddr;
+
+	devmem = open("/dev/mem", O_RDWR | O_SYNC);
+	offset = (off_t)hw_addr % getpagesize();
+	page_addr = (off_t)(hw_addr - offset);
+
+	hw_addr = (intptr_t *)mmap(0, 4, PROT_READ|PROT_WRITE,
```
Comment (etienne-lms) [inline libteec/src/tee_bench_aux.c]: should check that `mmap()` succeeded.

File: libteec/src/tee_client_api.c [hunk 4]
```diff
@@ -553,6 +552,11 @@ TEEC_Result TEEC_InvokeCommand(TEEC_Session *session, uint32_t cmd_id,
 		goto out;
 	}
 
+	if ((cmd_id & BENCHMARK_CMD(0)) != BENCHMARK_CMD(0)) {
```
Comment (etienne-lms) [inline libteec/src/tee_client_api.c]: This expects the 0xFA19xxxx IDs are reserved to the benchmark pseudo TA. This is not fair.

File: public/tee_bench.h [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+#define TEE_BENCH_CLIENT	0x10000000
+#define TEE_BENCH_KMOD		0x20000000
+#define TEE_BENCH_CORE		0x30000000
+#define TEE_BENCH_UTEE		0x40000000
+#define TEE_BENCH_DUMB_TA	0xF0000001
+
+struct statistics {
+	int n;
+	double m;
+	double M2;
```
Comment (etienne-lms) [inline public/tee_bench.h]: minor: i would rather reserve the capital letters to macros.

File: public/tee_bench.h [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+	struct tee_time_st stamps[];
+};
+
+#ifdef CFG_TEE_BENCHMARK
+/* Reading program counter */
+static inline __attribute__((always_inline)) uintptr_t read_pc(void)
+{
+	uintptr_t pc;
+#ifdef __aarch64__
+	asm volatile("adr %0, ." : "=r"(pc));
```
Comment (etienne-lms) [inline public/tee_bench.h]: `asm volatile("adr %0, ." : "=r"(pc));` should work for both AArch32 and AArch64.

File: libteec/src/tee_client_api.c (hunk 4 above)
Comment (igoropaniuk) [inline libteec/src/tee_client_api.c]: Right, but the only one alternative(IMHO) is to hijack(how it's done for benchmark buf) session_id (opened to Benchmark PTA) and store it in all OP_TEE layers
If it's definitely better solution, I'll use it

**Comment by etienne-lms** [comment]
This PR seems to deprecate https://github.com/OP-TEE/optee_client/pull/63. If so, could you close it ?

File: libteec/src/tee_bench_aux.c (hunk 1 above)
Comment (igoropaniuk) [inline libteec/src/tee_bench_aux.c]: fixed

File: libteec/src/tee_bench_aux.c (hunk 2 above)
Comment (igoropaniuk) [inline libteec/src/tee_bench_aux.c]: removed all statistics handling code

File: libteec/src/tee_client_api.c (hunk 4 above)
Comment (igoropaniuk) [inline libteec/src/tee_client_api.c]: fixed. Now timestamp buffer is registered via RPC call from SW

File: public/tee_bench.h (hunk 5 above)
Comment (igoropaniuk) [inline public/tee_bench.h]: fixed

File: libteec/include/linux/teec_benchmark.h [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef __TEEC_BENCHMARK_H
+#define __TEEC_BENCHMARK_H
+
+#include <linux/types.h>
+#include <stdbool.h>
```
Comment (jbech-linaro) [inline libteec/include/linux/teec_benchmark.h]: Is this needed here?

File: libteec/include/linux/teec_benchmark.h [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+#include <stdio.h>
+#include <sys/types.h>
```
Comment (jbech-linaro) [inline libteec/include/linux/teec_benchmark.h]: Same here

File: libteec/src/teec_benchmark.c [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+	uint64_t ccounter = 0;
+#ifdef __aarch64__
+	asm volatile("mrs %0, PMCCNTR_EL0" : "=r"(ccounter));
+#else
+	asm volatile("mrc p15, 0, %0, c9, c13, 0" : "=r"(ccounter));
+#endif
+	return ccounter * TEE_BENCH_DIVIDER;
+}
+
+/* Misc auxilary functions */
```
Comment (jbech-linaro) [inline libteec/src/teec_benchmark.c]: Doesn't matter now, but in the long run we'd better move it into teec_trace.c  (and cleanup that file. I don't think the stuff in there is used any longer ... old ST-Ericsson code).

File: libteec/src/teec_benchmark.c [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+	TEEC_FinalizeContext(&bench_ctx);
+}
+
+static void benchmark_get_bench_buf_paddr(uint64_t *paddr_ts_buf)
+{
+	TEEC_Result res;
+	TEEC_Operation op = { 0 };
+	uint32_t ret_orig;
+
+	benchmark_pta_open();
```
Comment (etienne-lms) [inline libteec/src/teec_benchmark.c]: If this fails, the command should not be invoked.

File: libteec/src/teec_benchmark.c [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+	op.paramTypes = TEEC_PARAM_TYPES(TEEC_VALUE_OUTPUT, TEEC_NONE,
+					TEEC_NONE, TEEC_NONE);
+
+	res = TEEC_InvokeCommand(&bench_sess, BENCHMARK_CMD_GET_MEMREF,
+					&op, &ret_orig);
+	tee_check_res(res, "TEEC_InvokeCommand");
```
Comment (etienne-lms) [inline libteec/src/teec_benchmark.c]: If this fails, `*paddr_ts_buf` should not be filled.

File: libteec/src/teec_benchmark.c [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+
+	benchmark_pta_close();
+}
+
+static void *mmap_paddr(intptr_t paddr)
+{
+	int   devmem;
+	off_t offset = 0;
+	off_t page_addr;
+
```
Comment (etienne-lms) [inline libteec/src/teec_benchmark.c]: minor: empty line

File: libteec/src/teec_benchmark.c [hunk 13]
```diff
@@ -0,0 +1,211 @@
...
+	intptr_t *hw_addr = (intptr_t *)paddr;
+
+	devmem = open("/dev/mem", O_RDWR | O_SYNC);
+	offset = (off_t)hw_addr % getpagesize();
+	page_addr = (off_t)(hw_addr - offset);
+
+	hw_addr = (intptr_t *)mmap(0, 4, PROT_READ|PROT_WRITE,
+					MAP_SHARED, devmem, page_addr);
+	if (hw_addr == MAP_FAILED)
+		return 0;
```
Comment (etienne-lms) [inline libteec/src/teec_benchmark.c]: minor: `return NULL`

File: libteec/src/teec_benchmark.c [hunk 14]
```diff
@@ -0,0 +1,211 @@
...
+
+	return (hw_addr + offset);
```
Comment (etienne-lms) [inline libteec/src/teec_benchmark.c]: minor: remove parenthesis. (note: label "hw_addr" is a bit confusing).

File: libteec/src/teec_benchmark.c [hunk 15]
```diff
@@ -0,0 +1,211 @@
...
+{
+	uint64_t ts_buf_raw = 0;
+	bool res = true;
+
+	if (!bench_ts_global) {
+		/* receive buffer from Benchmark PTA and register it */
+		benchmark_get_bench_buf_paddr(&ts_buf_raw);
+		if (ts_buf_raw) {
+			bench_ts_global = mmap_paddr(ts_buf_raw);
+			res = true;
```
Comment (etienne-lms) [inline libteec/src/teec_benchmark.c]: Check `bench_ts_global != NULL` before returning `true`.

File: libteec/src/teec_benchmark.c [hunk 16]
```diff
@@ -0,0 +1,211 @@
...
+		} else {
+			res = false;
+		}
+	}
+
+	return res;
+}
+
+void bm_init_ts_buf(void)
```
Comment (etienne-lms) [inline libteec/src/teec_benchmark.c]: do we really need both `bm_init_ts_buf()` and `benchmark_check_mode()` ?

File: libteec/src/teec_benchmark.c [hunk 17]
```diff
@@ -0,0 +1,211 @@
...
+{
+	benchmark_check_mode();
+}
+
+/* Adding timestamp to buffer */
+void bm_timestamp(void)
+{
+	cpu_set_t cpu_set_old, cpu_set_tmp;
```
Comment (etienne-lms) [inline libteec/src/teec_benchmark.c]: minor: split in 2 lines.

File: libteec/src/teec_benchmark.c [hunk 18]
```diff
@@ -0,0 +1,211 @@
...
+
+	cpu_buf = &bench_ts_global->cpu_buf[cur_cpu];
+	ts_i = __sync_fetch_and_add(&cpu_buf->head, 1);
+	ts_data.cnt = read_ccounter();
+	ts_data.addr = (uintptr_t) ret_addr;
+	ts_data.src = TEE_BENCH_CLIENT;
+	cpu_buf->stamps[ts_i & TEE_BENCH_MAX_MASK] = ts_data;
+
+	/* set back affinity mask */
+	ret = sched_setaffinity(0, sizeof(cpu_set_old), &cpu_set_old);
```
Comment (etienne-lms) [inline libteec/src/teec_benchmark.c]: minor: drop return value.

File: public/tee_bench.h [hunk 19]
```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright (c) 2017, Linaro Limited
+ *
+ * This software is licensed under the terms of the GNU General Public
```
Comment (etienne-lms) [inline public/tee_bench.h]: Prefer the BSD style licence.

File: public/tee_bench.h [hunk 20]
```diff
@@ -0,0 +1,58 @@
...
+#include <inttypes.h>
+
+#define UNUSED(x) (void)(x)
+
+#define BENCHMARK_CMD(id)	(0xFA190000 | ((id) & 0xFFFF))
+#define BENCHMARK_CMD_REGISTER_MEMREF	BENCHMARK_CMD(1)
+#define BENCHMARK_CMD_GET_MEMREF		BENCHMARK_CMD(2)
+#define BENCHMARK_CMD_UNREGISTER		BENCHMARK_CMD(3)
+
+#define TEE_BENCH_DIVIDER		64
```
Comment (etienne-lms) [inline public/tee_bench.h]: Add same comment as in **optee_os/core/.../bench.h**:
```c
/*
 * Cycle count divider is enabled (in PMCR), 
 * CCNT value is incremented every 64th clock cycle
 */
```

File: libteec/src/teec_benchmark.c [hunk 21]
```diff
@@ -0,0 +1,211 @@
...
+	off_t offset = 0;
+	off_t page_addr;
+
+	intptr_t *hw_addr = (intptr_t *)paddr;
+
+	devmem = open("/dev/mem", O_RDWR | O_SYNC);
+	offset = (off_t)hw_addr % getpagesize();
+	page_addr = (off_t)(hw_addr - offset);
+
+	hw_addr = (intptr_t *)mmap(0, 4, PROT_READ|PROT_WRITE,
```
Comment (etienne-lms) [inline libteec/src/teec_benchmark.c]: This request maps of only 4 bytes. Should better map up to `cores` field, then remap the full bench ts data structure.

File: libteec/src/teec_benchmark.c [hunk 22]
```diff
@@ -0,0 +1,211 @@
...
+
+static void *mmap_paddr(intptr_t paddr)
+{
+	int   devmem;
+	off_t offset = 0;
+	off_t page_addr;
+
+	intptr_t *hw_addr = (intptr_t *)paddr;
+
+	devmem = open("/dev/mem", O_RDWR | O_SYNC);
```
Comment (etienne-lms) [inline libteec/src/teec_benchmark.c]: Nicely warn and exit if this fails. Maybe /dev/mem is not accessible.

File: libteec/src/teec_benchmark.c [hunk 23]
```diff
@@ -95,41 +84,51 @@ static void benchmark_pta_close(void)
...
+		return res;
+
 	op.paramTypes = TEEC_PARAM_TYPES(TEEC_VALUE_OUTPUT, TEEC_NONE,
 					TEEC_NONE, TEEC_NONE);
 
 	res = TEEC_InvokeCommand(&bench_sess, BENCHMARK_CMD_GET_MEMREF,
 					&op, &ret_orig);
-	tee_check_res(res, "TEEC_InvokeCommand");
+	if (res != TEEC_SUCCESS)
+		return res;
```
Comment (etienne-lms) [inline libteec/src/teec_benchmark.c]: `benchmark_pta_close();`

File: libteec/include/linux/teec_benchmark.h [hunk 24]
```diff
@@ -0,0 +1,37 @@
...
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef __TEEC_BENCHMARK_H
+#define __TEEC_BENCHMARK_H
+
+#include <stdbool.h>
+#include <tee_bench.h>
+
+void bm_init_ts_buf(void);
```
Comment (etienne-lms) [inline libteec/include/linux/teec_benchmark.h]: minor: deprecated `bm_init_ts_buf()`

File: libteec/src/teec_benchmark.c [hunk 25]
```diff
@@ -0,0 +1,208 @@
...
+	ts_data.cnt = read_ccounter();
+	ts_data.addr = (uintptr_t) ret_addr;
+	ts_data.src = TEE_BENCH_CLIENT;
+	cpu_buf->stamps[ts_i & TEE_BENCH_MAX_MASK] = ts_data;
+
+	/* set back affinity mask */
+	sched_setaffinity(0, sizeof(cpu_set_old), &cpu_set_old);
+}
+#else /* CFG_TEE_BENCHMARK */
+void bm_init_ts_buf(void) {}
```
Comment (etienne-lms) [inline libteec/src/teec_benchmark.c]: minor: deprecated `bm_init_ts_buf()`

File: libteec/src/tee_client_api.c [hunk 26]
```diff
@@ -553,6 +554,14 @@ TEEC_Result TEEC_InvokeCommand(TEEC_Session *session, uint32_t cmd_id,
 		goto out;
 	}
 
+#ifdef CFG_TEE_BENCHMARK
+	if (!pthread_mutex_trylock(&teec_bench_mu)) {
```
Comment (etienne-lms) [inline libteec/src/tee_client_api.c]: If the lock is held, this will not get any timestamp. Is this expected ? (maybe this has already been discussed)

Comment (etienne-lms) [inline libteec/src/tee_client_api.c]: `The full sequence (including lock and get-buffer) could be implemented inside `bm_timestamp()`.

File: libteec/src/teec_benchmark.c [hunk 27]
```diff
@@ -0,0 +1,208 @@
...
+
+	devmem = open("/dev/mem", O_RDWR | O_SYNC);
+	if (!devmem)
+		return NULL;
+
+	offset = (off_t)hw_addr % getpagesize();
+	page_addr = (off_t)(hw_addr - offset);
+
+	hw_addr = (intptr_t *)mmap(0, size, PROT_READ|PROT_WRITE,
+					MAP_SHARED, devmem, page_addr);
```
Comment (etienne-lms) [inline libteec/src/teec_benchmark.c]: Can (should) `close(devmem);` from here (after `mmap()` returns).

File: libteec/src/tee_client_api.c (hunk 26 above)
Comment (igoropaniuk) [inline libteec/src/tee_client_api.c]: fixed

File: libteec/include/linux/teec_benchmark.h (hunk 24 above)
Comment (igoropaniuk) [inline libteec/include/linux/teec_benchmark.h]: fixed

File: libteec/src/teec_benchmark.c (hunk 25 above)
Comment (igoropaniuk) [inline libteec/src/teec_benchmark.c]: fixed

File: libteec/src/teec_benchmark.c (hunk 27 above)
Comment (igoropaniuk) [inline libteec/src/teec_benchmark.c]: fixed

**Comment by igoropaniuk** [comment]
Rebased on latest master and tested

File: libteec/src/teec_benchmark.c [hunk 28]
```diff
@@ -0,0 +1,218 @@
...
+	/* fill timestamp data */
+	if (cur_cpu >= bench_ts_global->cores) {
+		ret = sched_setaffinity(0, sizeof(cpu_set_old), &cpu_set_old);
+		goto error;
+	}
+
+	ret_addr = __builtin_return_address(0);
+
+	cpu_buf = &bench_ts_global->cpu_buf[cur_cpu];
+	ts_i = __sync_fetch_and_add(&cpu_buf->head, 1);
```
Comment (igoropaniuk) [inline libteec/src/teec_benchmark.c]: @etienne-lms why useless?
we can be preempted here

Comment (etienne-lms) [inline libteec/src/teec_benchmark.c]: your lock + cpu affinity prevent another thread from accessing the buffer.

Comment (igoropaniuk) [inline libteec/src/teec_benchmark.c]: `sched_setaffinity` call just tells linux scheduler that this thread should be scheduled on specified core (or cores), it doesn't somehow affect preemption.

It's theoretically possible, that while running simple non-atomic `cpu_buf->head++` thread will be preempted, and another one will use non-actual `cpu-buf->head` value. Or maybe I just didn't get what actually you meant

Comment (igoropaniuk) [inline libteec/src/teec_benchmark.c]: @etienne-lms

Comment (etienne-lms) [inline libteec/src/teec_benchmark.c]: Ok, understood. Thanks.

File: Makefile [hunk 29]
```diff
@@ -29,7 +29,9 @@ all: build install
 
 build-libteec:
 	@echo "Building libteec.so"
-	@$(MAKE) --directory=libteec --no-print-directory --no-builtin-variables
+	@$(MAKE) --directory=libteec CFG_TEE_BENCHMARK=$(CFG_TEE_BENCHMARK) \
+	CFG_TEE_PRINT_LATENCY_STAT=$(CFG_TEE_PRINT_LATENCY_STAT) \
+	--no-print-directory --no-builtin-variables
```
Comment (etienne-lms) [inline Makefile]: minor style: maybe insert an indentation in the 2 later line.

File: Makefile [hunk 30]
```diff
@@ -29,7 +29,9 @@ all: build install
 
 build-libteec:
 	@echo "Building libteec.so"
-	@$(MAKE) --directory=libteec --no-print-directory --no-builtin-variables
+	@$(MAKE) --directory=libteec CFG_TEE_BENCHMARK=$(CFG_TEE_BENCHMARK) \
+	CFG_TEE_PRINT_LATENCY_STAT=$(CFG_TEE_PRINT_LATENCY_STAT) \
```
Comment (etienne-lms) [inline Makefile]: `CFG_TEE_PRINT_LATENCY_STAT` seems not used. Maybe discard it until some code rely on its value.

File: Makefile (hunk 29 above)
Comment (igoropaniuk) [inline Makefile]: fixed

File: Makefile (hunk 30 above)
Comment (igoropaniuk) [inline Makefile]: fixed

**Comment by igoropaniuk** [comment]
Rebased & applied tags

**Pull request merged**